evm/p2p/lightchain.py

```diff
@@ -1,69 +1,270 @@
+                self.logger.error("Unexpected error (%s), restarting", traceback.format_exc())
+                await self.stop_all_peers()
+            # Wait self._connect_loop_sleep seconds, unless we're asked to stop.
+            await asyncio.wait([self._should_stop.wait()], timeout=self._connect_loop_sleep)
+        self._finished.set()
+
+    async def stop_all_peers(self):
+        self.logger.info("Stopping all peers ...")
+        await asyncio.gather(
+            *[peer.stop_and_wait_until_finished() for peer in self.connected_nodes.values()])
 
     async def stop(self):
-        raise NotImplementedError("TODO")
+        self._should_stop.set()
+        await self.stop_all_peers()
+        await self._finished.wait()
+
+    async def connect(self, remote):
+        """
+        Connect to the given remote and return a Peer instance when successful.
+
+        Returns None if the remote is unreachable, times out or is useless.
+        """
+        if remote in self.connected_nodes:
+            self.logger.debug("Skipping %s; already connected to it", remote)
+            return None
+        try:
+            peer = await asyncio.wait_for(
+                handshake(remote, self.privkey, self.peer_class, self.chaindb, self.network_id),
+                HANDSHAKE_TIMEOUT)
+            return peer
+        except (UnreachablePeer, asyncio.TimeoutError, PeerConnectionLost) as e:
+            self.logger.debug("Could not complete handshake with %s: %s", remote, e)
+        except UselessPeer:
+            self.logger.debug("No matching capabilities with %s", remote)
+        except PeerDisconnected as e:
+            self.logger.debug(
+                "%s disconnected before completing handshake; reason: %s", remote, e)
+        except Exception:
+            self.logger.warn(
+                "Unexpected error during auth/p2p handhsake with %s: %s",
+                remote, traceback.format_exc())
+        return None
+
+    async def maybe_connect_to_more_peers(self):
+        """Connect to more peers if we're not yet connected to at least self.min_peers."""
+        if len(self.connected_nodes) >= self.min_peers:
+            self.logger.debug(
+                "Already connected to %s peers: %s; sleeping",
+                len(self.connected_nodes),
+                [remote for remote in self.connected_nodes])
+            return
+
+        for node in await self.get_nodes_to_connect():
+            self.logger.debug("Connecting to %s", node)
+            peer = await self.connect(node)
+            if peer is not None:
+                self.logger.info("Successfully connected to %s", peer)
+                asyncio.ensure_future(peer.start(finished_callbabck=self.peer_finished))
+                self.connected_nodes[peer.remote] = peer
+
+    def peer_finished(self, peer):
+        self.logger.info("Peer finished: %s", peer)
+        if peer.remote in self.connected_nodes:
```

Comment:
This code has me confused.  Here we pop the peer out of the `connected_nodes` but above on line 126, we immediately place it into the `connected_nodes` once this finishes.  Not intuitive to me why this code does what it does.  Can you maybe add a comment?

Reply:
This method is passed as the finished_callback arg to peer.start(). It is only called when the peer disconnects and peer.start() returns. I'd be happy to add a comment but not sure what it should say

Reply:
Maybe making it internal `_peer_finished()` makes it clearer that it's not meant to be useful on its own.

As for my confusion, I think I understand now.  This is *queued* to be run later.  Line 125 puts the remote in the `connected_nodes` list and then this removes it later when it's finished.  Maybe if line 124 and 125 were swapped this would be more intuitive?

Reply:
Sure, I've done that and also added a docstring to `_peer_finished()`